### DISCUSSION TITLE
Docker support

### DISCUSSION THREAD

**Discussion by janc18** [afpineda/OpenSourceSimWheelESP32#17] [Ideas]
Hi @afpineda,

I was thinking that, to have a better platform for testing new builds, we could use Docker to test the compilation results. This would make it easier to detect when certain dependencies break something. Currently, I can compile the Arduino projects using arduino-cli, but I need to make some changes to the Dockerfile to simplify the compilation process.

I tried to create a pull request under the development branch but couldn't, and I also faced issues creating a new branch. If you give me permission, I'd be happy to upload my pull request. (I forgot to make a pull request from my fork 😆 )

**Comment by afpineda** [comment]
I want to keep this project simple and accessible to non-IT people, so Docker is definitely out. However, I share your concerns:

**About dependencies**

> This would make it easier to detect when certain dependencies break something.

This project keeps dependencies to an absolute minimum:

- Development dependencies:
  - Powershell/Bash (In fact, I thought that PowerShell was available for Linux.).
  - Arduino IDE/cli

  (Note that Docker would introduce another development dependency)

- Library dependencies:
  - NimBLE-Arduino
  - ESP-Arduino

This project deliberately avoids Arduino bloatware like `analogRead()` and the like. 
Most of the code relies on the underlying ESP-IDF API with the exception of the `Preferences` class.
Even the LED-strip code does not rely on external libraries or ESP-Arduino.
This is a design principle, even if not explicitly stated.

**About Continuous Integration**

Automatic compilation using Arduino-CLI was on my mind, so feel free to contribute in that sense. A simple script should do it:

- Compile each unit/integration test
- Compile each "SetupXX.ino" sketch
- Dump output to a file for review.


However, each sketch must target a different board.
I have no knowledge of Arduino CLI, so I avoided this.

I have to admit that manual compilation is tedious and time-consuming. 
But since most tests are not automated and require specific hardware, automatic compilation does not improve development too much.

**Pull-request**

> I tried to create a pull request under the development branch but couldn't, and I also faced issues creating a new branch. If you give me permission, I'd be happy to upload my pull request.

You should be able to target the `development` branch for a pull-request.
In fact, you already [did](https://github.com/afpineda/OpenSourceSimWheelESP32/pull/18).

**Reply by janc18** [reply]
I understand that using Docker might be a bit challenging to set up, but as you mentioned:

> About Continuous Integration
> Automatic compilation using Arduino-CLI was on my mind, so feel free to contribute in that regard. A simple script should do it:
> Compile each unit/integration test
> Compile each 'SetupXX.ino' sketch
> Dump output to a file for review.

It could be a good internal tool that can be executed with GitHub Actions, which would provide the output and display it in the README.
Let me test how it goes in my fork.

**Reply by afpineda** [reply]
I added batch compilation using `arduino-cli`. 
See [this commit](https://github.com/afpineda/OpenSourceSimWheelESP32/commit/371dac1ba7c489152193df2d07889be438bf3ebe).

**Comment by janc18** [comment]
Hi @afpineda  i added a workflow at my repository, tomorrow i will make a PR to Development to include the github actions, 
Check the output
![Github actions output](https://github.com/janc18/OpenSourceSimWheelESP32/actions/runs/11849654390/job/33023270007) 
And in the future it could display the result directly at the README file.

**Reply by afpineda** [reply]
Something is not working:

- The Python virtual envirnoment was not activated:

```
944: ./autocompile.sh: 22: source: not found
```

- The strange compilation errors at "I2C_Probe" are reproduced at "Setup9": output lines 984 to 996.
- Virtual environment deactivation does not work:

```
1042: ./autocompile.sh: 51: deactivate: not found
```

Note that this script should compile without errors as Arduino IDE and BatchCompile.ps1 do.

**Reply by afpineda** [reply]
This is the output from "BatchCompile.ps1", run today:
[compile.txt](https://github.com/user-attachments/files/17771367/compile.txt)

I have no spare time right now to help more with this.
In a couple of weeks I will go deeper into GitHub actions. I promise.
In fact, automating D'oxygen documentation was in my mind, too.

**Reply by janc18** [reply]
Ok let me fix the error related to I2c_tools and I could start checking the D'oxygen documentation generation

**Reply by afpineda** [reply]
D'oxygen is not a priority. There is a reusable Action in the GitHub marketplace for that. It should be easy.

**Reply by afpineda** [reply]
The Doxygen stuff is finished.
Deploys to [https://afpineda.github.io/OpenSourceSimWheelESP32/](https://afpineda.github.io/OpenSourceSimWheelESP32/)
It was harder than expected due to environment protection rules.
See [APIdoc.yaml](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/main/.github/workflows/APIdoc.yaml)

**Comment by afpineda** [comment]
Hi, @janc18.

The build workflow is **updated and working** in my `development` branch.
Please, synchronize your cloned repo.
It will be available after merging `development` into `main`.
This was tested in a separate private repo.

See [build.yaml](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/Development/.github/workflows/build.yaml)
and [autocompile.sh](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/Development/src/autocompile.sh)

**Reply by janc18** [reply]
Hi @afpineda, awesome, I read the scripts and you added optimizations, I will update mi fork today
